pageserver/src/tenant.rs

```diff
@@ -1354,36 +1355,32 @@ impl Tenant {
                     }
                 }
 
-                // Ideally we should use Tenant::set_broken_no_wait, but it is not supposed to be used when tenant is in loading state.
-                enum BrokenVerbosity {
-                    Error,
-                    Info
-                }
-                let make_broken =
-                    |t: &Tenant, err: anyhow::Error, verbosity: BrokenVerbosity| {
-                        match verbosity {
-                            BrokenVerbosity::Info => {
-                                info!("attach cancelled, setting tenant state to Broken: {err}");
-                            },
-                            BrokenVerbosity::Error => {
-                                error!("attach failed, setting tenant state to Broken: {err:?}");
-                            }
-                        }
-                        t.state.send_modify(|state| {
-                            // The Stopping case is for when we have passed control on to DeleteTenantFlow:
-                            // if it errors, we will call make_broken when tenant is already in Stopping.
-                            assert!(
-                                matches!(*state, TenantState::Attaching | TenantState::Stopping { .. }),
-                                "the attach task owns the tenant state until activation is complete"
-                            );
-
-                            *state = TenantState::broken_from_reason(err.to_string());
-                        });
+                // Ideally we should use Tenant::set_broken_no_wait, but it is not supposed to be
+                // used when tenant is in loading state.
+                fn make_broken_or_stopping(t: &Tenant, err: anyhow::Error) {
+                    // If the tenant is cancelled, assume the error was caused by cancellation.
+                    let new_state = if t.cancel.is_cancelled() {
+                        info!("attach cancelled, setting tenant state to Stopping: {err}");
+                        TenantState::Stopping { progress: Barrier::default() }
```

Comment:
We may also have to set the timelines to `Stopping` here, similarly to `set_stopping()`. Otherwise, we may already have activated some timelines before we're cancelled.

Does that sound right @jcsp? Or doesn't that matter since the tenant hasn't fully activated yet?

Reply:
Hmm.  We shouldn't really need to recurse into timelines here because shutdown() already does that, and we know it's running because it set the cancellation token.

However, looking at Tenant::shutdown(), I see that it has a special behavior for if the tenant is already stopping, where it drops out early.

So if we set stopping here, but don't clean up timelines, then shutdown() isn't going to bother shutting down timelines :-/

I wonder what happens if we just leave the tenant in state Attaching rather than going to Broken?  The page_service behavior would still be to block access to the tenant, and shutdown would still understand that it needs to tear down timelines

Reply:
Not trivially, because `set_stopping` waits for the tenant to exit `Attaching` state. I believe that was the motivation for setting it to `Broken` in the first place.

https://github.com/neondatabase/neon/blob/39cb64e653a37eeaaa92331dbaa51428778d1807/pageserver/src/tenant.rs#L3520-L3531

Reply:
A few possible options:

1. Restructure shutdown such that it handles `attach` setting `ShuttingDown`.
2. Make `set_stopping` idempotent such that multiple callers can call it, and have `attach` call it too.
3. Add a separate `TenantState::AttachCancelled` state.
4. Add a `TenantState::Broken` field indicating attach was cancelled.

I'm inclined to try 1 to reduce the number of states.

Reply:
I don't think we can easily restructure the code to handle this. I instead made `Stopping::progress` an `Option`, and use the `None` state to signal attach cancellation to `set_stopping`.

I'm not thrilled about this, but I think it gets the job done in a fairly targeted and low-risk manner. Wdyt?

Reply:
Not beautiful but it's a net improvement to avoid 500s

Reply:
Tests all passed, this should be ready for review.

I added a testing assertion to check if the `TenantState::Stopping` case during attach is still valid.

Will rebase #11328 on top of this to verify that tests pass there too.